### PR TITLE
feat!: auto extract exports

### LINKED ISSUE #103: Auto extract package exports
We can directly use the npm package to automatically detect their exports and use it as a preset. Currently, it is only possible to use manual presets and a list of their exports.

In order to extract exports, we have 3 possible ways:
- Actually require package in an isolated runtime and check the exports
- Use static syntax analyses to detect exports 
- Use metadata directly exported from packages

Using the last method is safest because packages can exactly annotate their exports and use a built step plugin (unbuild or rollup) to automate this meta generation process into `dist/unimport.json`. We can then refer it from `package.json` using `"unimport": "./dist/unimport.json"`

When packages are not bundled in this way, we can fallback to [mlly.resolveModuleExportNames](https://github.com/unjs/mlly#resolvemoduleexportnames) and use fast static analyzes to extract export names.

In order to have better control over auto-infered exports, we might exprot a filter function from presets. I imagine preset looking like this:

```js
presets: [{
  package: 'h3',
  exports: 'auto' | 'analyze' // Default is auto - Tries `autoImport` in pacakge.json then analyzes
  ignore: ['h', (name) => !name.startsWith('use')]
}]
```

In order to speed up the performance, we might use global cache in `{os.tmpDir}/unimport/{package}/{version}/exports.json` containing exports and an integrity field to validate cache entry. 

(This is a summary of discussing together with @antfu)

### PR DESCRIPTION
**Pull request by antfu**
resolve #103

Todo:
- [x] Support `PackagePreset`
- [x] Global Cache
- [ ] Package.json's `meta.unimport` field support

Breaking changes:

- Internal context type change
- `generateTypeDeclarations`, `getExports` become async

### CODE REVIEW

File: src/extract.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+
+  return filtered.map(name => (<Import>{
+    from: preset.package,
+    name
+  }))
+}
+async function extractExports (name: string, url?: string, cache = true) {
+  const packageJsonPath = await resolvePackageJSON(name, { url })
+  const packageJson = await readPackageJSON(packageJsonPath)
+  const version = packageJson.version
```
Comment (pi0) [inline src/extract.ts:34]: This is rare and shouldn't happen but private packages can omit version 
```suggestion
  const version = packageJson.version ||  '0.0.0'
```

File: src/extract.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+  const cachePath = join(os.tmpdir(), 'unimport', name + '@' + version, 'exports.json')
+
+  const hasCache = cache && existsSync(cachePath)
```
Comment (pi0) [inline src/extract.ts:37]: Some ideas from jiti (src: https://github.com/unjs/jiti/blob/e44a2679cc3f3f529fbd81668e51bedcd20b74f9/src/jiti.ts#L72)

In some deployment platforms, tmp dir can be not writable

File: src/extract.ts (hunk 1 above)
Comment (antfu) [reply src/extract.ts:46]: Updated the impl, I guess it's better to skip the cache for private packages as they are addressed to change right?

**Comment by antfu** [comment]
I guess we could have it for now, and leave `meta.unimport` for another PR with a more detailed design?

**Review by pi0 — APPROVED** [review]

**Pull request merged**